### PR TITLE
SNOW-1568022: [Local Testing] Add support for within_group and array_agg

### PR DESCRIPTION
**Pull request by sfc-gh-jrose**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1568022

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

   This PR implements the array_agg function and adds support for WithinGroup clauses.

### CODE REVIEW

File: src/snowflake/snowpark/mock/_functions.py [hunk 1]
```diff
@@ -390,6 +390,15 @@ def mock_covar_pop(column1: ColumnEmulator, column2: ColumnEmulator) -> ColumnEm
     )
 
 
+@patch("array_agg")
+def mock_array_agg(column: ColumnEmulator, is_distinct: bool) -> ColumnEmulator:
+    columns_data = ColumnEmulator(column.unique()) if is_distinct else column
+    return ColumnEmulator(
+        data=[list(columns_data.dropna())],
+        sf_type=ColumnType(ArrayType(StringType()), False),
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/mock/_functions.py:398]: dumb question, why it's StringType not the data type of the input column?

File: src/snowflake/snowpark/mock/_plan.py [hunk 2]
```diff
@@ -259,6 +260,13 @@ def __init__(
         self.options = options
 
 
+def coerce_order_by_arguments(order_by: List[Union[Expression]]):
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/mock/_plan.py:263]: nit: for type hint, is it missing other types for Union?

File: tests/integ/scala/test_dataframe_aggregate_suite.py [hunk 3]
```diff
@@ -1336,7 +1332,7 @@ def test_listagg_within_group(session):
     )
     Utils.check_answer(
         df.group_by("color")
-        .agg(listagg("length", ",").within_group(df.length.asc()))
```
Comment (sfc-gh-aling) [inline tests/integ/scala/test_dataframe_aggregate_suite.py:1339]: just curious why we did this change, and are they equivalent?

File: src/snowflake/snowpark/mock/_functions.py (hunk 1 above)
Comment (sfc-gh-jrose) [reply src/snowflake/snowpark/mock/_functions.py:398]: I can actually probably remove that. The old default was for it to be string type, but I think with some of my more recent changes it should be able to adapt without specifying. That way when structured type support is done rolling out it won't need to be updated her.

File: tests/integ/scala/test_dataframe_aggregate_suite.py (hunk 3 above)
Comment (sfc-gh-jrose) [reply tests/integ/scala/test_dataframe_aggregate_suite.py:1339]: The default sorting order is ascending so I wanted to make sure that was the case with this change. These two statements are equivalent in live mode.

**Review by sfc-gh-aling — APPROVED** [review]

**Pull request merged**